src/neural/cuda/common_kernels.cu

```diff
@@ -493,6 +493,572 @@ void PolicyMap(int N, T* output, const T* input, const short* indices,
   ReportCUDAErrors(cudaGetLastError());
 }
 
+namespace ConvKernel0 {
+// Special kernel to get better performance for small batch size (N=1)
+//
+// Optimizations:
+// - Do multiple elements in H and W dimensions (4x4) per thread, to get some
+//   spatial reuse for input tensor as well as filter.
+// - More spatial reuse for input tensor using shfl for reading elements outside
+//   the 4x4 'tile'.
+// - The sum across C dimension is split across multiple threads (8), and
+//    summed in parallel using SHFL.
+
+// Do multiple elements in H and W dimensions (4x4).
+constexpr int wPerThread = 4;
+constexpr int hPerThread = 4;
+
+// 32 threads in block.
+// One board (4 threads, with 16 elements per thread).
+constexpr int blockWidth = 4; 
+
+// different 'C' dimensions
+constexpr int blockHeight = 8;
+
+#define INDEX_NCHW(n, c, h, w) ((n)*C * 64 + (c)*64 + (h)*8 + w)
```

Comment:
I think in modern C++ it's better to do with constexpr functions (if types of arguments are known at least).